### ISSUE TITLE
IPython kernel autocompletions (some) stop working after installing LSP 

### ISSUE THREAD

**Issue by jpcaram** [jupyter-lsp/jupyterlab-lsp#1014]
## Description

Some of IPython kernel's autocompletions stop working after installing `jupyterlab-lsp`.

## Reproduce

1. In a fresh virtualenv, Install Jupyterlab and jupyterlab-lsp (`pip install jupyterlab-lsp`)
2. Start jupyterlab, create a notebook (Python kernel)
3. In a cell run:
```python
from IPython.core.latex_symbols import latex_symbols
latex_symbols['\\p'] = "PRINT!"
```
4. In another cell, type `\p` then hit `<tab>`. Nothing happens.
5. Shutdown Jupyter.
6. Uninstall the LSP (`pip uninstall jupyter-lsp jupyterlab-lsp`)
7. Repeat steps 2, 3 and 4. Autocomplete offers `PRINT!` as an option.

Parts of the kernel's autocompletion seem to work normally, like existing variable names and dictionary keys.

Also, `%config Completer.use_jedi = False` does not seem to have an effect. Installing a Python LSP server does not make a difference either.

## Expected behavior

The kernel's autocompletion is expected to work the same with or without jupyterlab-lsp installed.

## Context

- Operating System and version: `Linux jp-x1carbon 5.15.0-88-generic #98~20.04.1-Ubuntu SMP Mon Oct 9 16:43:45 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux`
- Browser and version: Firefox 119.0.1
- JupyterLab version: 4.0.8
- Language server and version: None

Other:
```
jupyter-client            8.6.0    
jupyter-core              5.5.0    
jupyter-events            0.9.0    
jupyter-lsp               2.2.0    
jupyter-server            2.10.1   
jupyter-server-terminals  0.4.4    
jupyterlab                4.0.8    
jupyterlab-lsp            5.0.0    
jupyterlab-pygments       0.2.2    
jupyterlab-server         2.25.1   
lsprotocol                2023.0.0
pygls                     1.2.0
pip                       20.0.2
```

<details><summary>Required: installed server extensions</summary>
<pre>
Config dir: /home/user/.jupyter

Config dir: /home/user/cloud/autocomplete/venv/etc/jupyter
    jupyter_lsp enabled
    - Validating jupyter_lsp...
Package jupyter_lsp took 0.0048s to import
A `_jupyter_server_extension_points` function was not found in jupyter_lsp. Instead, a `_jupyter_server_extension_paths` function was found and will be used for now. This function name will be deprecated in future releases of Jupyter Server.
      jupyter_lsp 2.2.0 OK
    jupyter_server_terminals enabled
    - Validating jupyter_server_terminals...
Package jupyter_server_terminals took 0.0022s to import
      jupyter_server_terminals 0.4.4 OK
    jupyterlab enabled
    - Validating jupyterlab...
Package jupyterlab took 0.0293s to import
      jupyterlab 4.0.8 OK
    notebook_shim enabled
    - Validating notebook_shim...
Package notebook_shim took 0.0000s to import
A `_jupyter_server_extension_points` function was not found in notebook_shim. Instead, a `_jupyter_server_extension_paths` function was found and will be used for now. This function name will be deprecated in future releases of Jupyter Server.
      notebook_shim  OK

Config dir: /usr/local/etc/jupyter

</pre>
</details>

<details><summary>Required: installed lab extensions</summary>
<pre>
Paste the output from running `jupyter labextension list` from the command line here.
You may want to sanitize the paths in the output.
</pre>
</details>

<details><summary>Troubleshoot Output</summary>
<pre>
JupyterLab v4.0.8
/home/user/autocomplete/venv/share/jupyter/labextensions
        jupyterlab_pygments v0.2.2 enabled  X (python, jupyterlab_pygments)
        @jupyter-lsp/jupyterlab-lsp v5.0.0 enabled OK (python, jupyterlab-lsp)

   The following extensions are outdated:
        jupyterlab_pygments
        
   Consider checking if an update is available for these packages.

Disabled extensions:
    @jupyterlab/completer-extension:base-service
    @jupyterlab/fileeditor-extension:language-server
    @jupyterlab/lsp-extension:settings
    @jupyterlab/notebook-extension:language-server
</pre>
</details>

<details><summary>Command Line Output</summary>
<pre>
Nothing obvious there.
</pre>
</details>

<details><summary>Browser Output (recommended for all interface issues)</summary>
<pre>
Nothing obvious there.
</pre>
</details>

**Comment by krassowski** [comment]
Thanks for the detailed report. Supporting this was a high priority and it used to work before the rewrite for JupyterLab 4.0. This is similar to #1009 and earlier issues on this topic.

<details>

<summary>This is a hard topic</summary>

The problem is that the kernels do not return the the information necessary to resolve whether the completion should be inserted after or in place of the existing text. This means that we need to use some heuristics for reconciliation of completions from various sources which breaks the edge cases like the latex symbols (which are substitutions). If someone has time to take a look I will prioritise review, but otherwise this might need a foundational change in jupyter protocol.

</details>

**Comment by jpcaram** [comment]
Thanks for your response.

To clarify, I used the `latex_symbols` examples just for conciseness. This appears to affect all custom completions. I will have to double check if those that are just completion and not replacement still work.

I would like to help, and I could use some information:
* Is there something implemented that is specifically disabling or ignoring certain kinds of completions?
* I tried to debug the issue at the kernel level by print statements in my custom completers. I don't see anything on the terminal. It appears that the custom completer's callback is never called. Or are kernel prints to stdout suppressed?
* Why would it be a foundational change to the protocol? I seems that this would just require additional fields in a message to indicate replacement/completion, position, etc.

Basically, I would like to understand the problem better.

Also, what is the alternative these days? Should one write their own LSP, add features to an existing one, write a Jupyter completer plugin?

**Comment by jpcaram** [comment]
Disabling "Pre-filter matches" solves the problem in my case. This may be an edge case though (Ipython kernel, Jupyter lab, Jupyter LSP support).

I still don't fully understand why this is a big issue from the perspective of Jupyter (server, lab, lsp). Things seem to work just fine with the IPython kernel. If the issue is for other kernels, it's an issue with their implementation, not Jupyter's. In my opinion, the option on the lsp client should be disabled by default. Otherwise, it will hide the problem for ever, and kernel developers won't address their problem.

Please correct me if I'm wrong on my assumptions.
